captum/attr/_utils/lrp_rules.py

```diff
@@ -0,0 +1,148 @@
+        output_hook = self._create_backward_hook_output(outputs.data)
+        self._handle_output_hook = outputs.register_hook(output_hook)
+
+    @staticmethod
+    def backward_hook_activation(module, grad_input, grad_output):
+        """Backward hook to propagate relevance over non-linear activations."""
+        return grad_output
+
+    def _create_backward_hook_input(self, inputs):
+        def _backward_hook_input(grad):
+            relevance = grad * inputs
+            self.relevance_input = relevance.data
+            return relevance
+
+        return _backward_hook_input
+
+    def _create_backward_hook_output(self, outputs):
+        def _backward_hook_output(grad):
+            sign = torch.sign(outputs)
+            sign[sign == 0] = 1
+            relevance = grad / (outputs + sign * self.STABILITY_FACTOR)
+            self.relevance_output = grad.data
+            return relevance
+
+        return _backward_hook_output
+
+    def forward_hook_weights(self, module, inputs, outputs):
+        """Save initial activations a_j before modules are changed"""
+        module.activations = tuple(input.data for input in inputs)
+        self._manipulate_weights(module, inputs, outputs)
+
+    @abstractmethod
+    def _manipulate_weights(self, module, inputs, outputs):
+        raise NotImplementedError
+
+    def forward_pre_hook_activations(self, module, inputs):
+        """Pass initial activations to graph generation pass"""
+        for input, activation in zip(inputs, module.activations):
+            input.data = activation
+        return inputs
+
+
+class EpsilonRule(PropagationRule):
+    """
+    Rule for relevance propagation using a small value of epsilon
+    to avoid numerical instabilities and remove noise.
+
+    Use for middle layers.
+
+    Args:
+        epsilon (integer, float): Value by which is added to the
+        discriminator during propagation.
+    """
+
+    def __init__(self, epsilon=1e-9):
+        self.STABILITY_FACTOR = epsilon
+
+    def _manipulate_weights(self, module, inputs, outputs):
+        pass
+
+
+class GammaRule(PropagationRule):
+    """
+    Gamma rule for relevance propagation, gives more importance to
+    positive relevance.
+
+    Use for lower layers.
+
+    Args:
+        gamma (float): The gamma parameter determines by how much
+        the positive relevance is increased.
+    """
+
+    def __init__(self, gamma=0.25, set_bias_to_zero=False):
+        self.gamma = gamma
+        self.set_bias_to_zero = set_bias_to_zero
+
+    def _manipulate_weights(self, module, inputs, outputs):
+        if hasattr(module, "weight"):
+            module.weight.data = (
+                module.weight.data + self.gamma * module.weight.data.clamp(min=0)
+            )
+        if self.set_bias_to_zero and hasattr(module, "bias"):
+            if module.bias is not None:
+                module.bias.data = torch.zeros_like(module.bias.data)
+
+
+class Alpha1_Beta0_Rule(PropagationRule):
```

Comment:
This is actually not the alpha beta rule but rather the z^+ rule, cf. appendix 1c in https://arxiv.org/abs/1512.02479
It only coincides with the alpha beta rule for alpha=1 iff all inputs are positive.